R/vroom_write.R

```diff
@@ -124,6 +124,12 @@ vroom_format <- function(x, delim = "\t", eol = "\n", na = "NA", col_names = TRU
                          bom = FALSE,
                          num_threads = vroom_threads()) {
 
+  stopifnot(is.data.frame(x))
+
+  if (length(x) == 0) {
```

Comment:
I think we need to consider the case where there are columns, but no rows. I guess we return the header string in that case? This scenario may need some analysis, i.e. if readr 1e does something in particular, what does base R do, etc.

``` r
(x <- data.frame(a = 1:2, b = 3:4))
#>   a b
#> 1 1 3
#> 2 2 4
x[0, ]
#> [1] a b
#> <0 rows> (or 0-length row.names)
length(x[0, ])
#> [1] 2
```

I think we also need to think carefully through the return value. For example, `character()` seems worth considering for the 0-columns case.

Reply:
In readr edition 1, `format_csv()` returns an empty string given an empty data frame. If the data frame only has column names, it returns those names. I'm not sure about base R yet. 


``` r
library(readr)

# empty data frame returns empty string

with_edition(1, format_csv(data.table::data.table()))
#> [1] ""

x <- data.frame(a = "", b = "")

x <- x[0, ]

# data frame with columns only returns column names

with_edition(1, format_csv(x))
#> [1] "a,b\n"
```

<sup>Created on 2022-04-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Reply:
OK I realize now that the 0-rows, >0-columns case was already covered, I just needed to ponder it.

As we just discussed, I think I'd change this from `length()` to `NCOL()` just on principle, to match what's done just above in `vroom_write()`.

Reply:
Oh gotta do this still!